### PR TITLE
Port VB RemoveUnusedMembers tests to Microsoft.CodeAnalysis.Testing f…

### PR DESCRIPTION
**Pull request by mavasani**
…ramework

### CODE REVIEW

File: src/EditorFeatures/VisualBasicTest/RemoveUnusedMembers/RemoveUnusedMembersTests.vb [hunk 1]
```diff
@@ -1330,46 +1400,48 @@ End Class")
...
 
     Public Sub M()
         Dim x = _goo4
     End Sub
 End Class",
 "Class C
     Private _goo4 As Char
-    Private _goo9, _goo10 As New String("""") ' Non constant initializer
-    Private _goo11 = 0  ' Implicit conversion to Object type in the initializer, hence it is a non constant initializer.
+    Private {|IDE0052:_goo9|}, {|IDE0052:_goo10|} As New String("""") ' Non constant initializer
```
Comment (mavasani) [inline src/EditorFeatures/VisualBasicTest/RemoveUnusedMembers/RemoveUnusedMembersTests.vb]: @sharwell This test is failing with the following message, and I believe it is a bug in the test framework:

```
Message: 
    Context: Fix all in document
    content of '/0/Test0.vb' did not match. Diff shown with expected as baseline:
     Class C
         Private _goo4 As Char
    -    Private _goo9, _goo10 As New String("") ' Non constant initializer
    -    Private _goo11 = 0  ' Implicit conversion to Object type in the initializer, hence it is a non constant initializer.
     
         Public Sub M()
             Dim x = _goo4
         End Sub
     End Class
```

If I removed the  `IDE0052` annotations in the result, the test fails with below message:
```
Message: 
    Microsoft.CodeAnalysis.Testing.Verifiers.EqualWithMessageException : Context: Diagnostics of fixed state
    Mismatch between number of diagnostics returned, expected "0" actual "3"
    
    Diagnostics:
    // /0/Test0.vb(3,13): info IDE0052: Private member 'C._goo9' can be removed as the value assigned to it is never read.
    VerifyVB.Diagnostic(VisualBasicRemoveUnusedMembersDiagnosticAnalyzer.IDE0052).WithSpan(3, 13, 3, 18),
    // /0/Test0.vb(3,20): info IDE0052: Private member 'C._goo10' can be removed as the value assigned to it is never read.
    VerifyVB.Diagnostic(VisualBasicRemoveUnusedMembersDiagnosticAnalyzer.IDE0052).WithSpan(3, 20, 3, 26),
    // /0/Test0.vb(4,13): info IDE0052: Private member 'C._goo11' can be removed as the value assigned to it is never read.
    VerifyVB.Diagnostic(VisualBasicRemoveUnusedMembersDiagnosticAnalyzer.IDE0052).WithSpan(4, 13, 4, 19),
```

File: src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs [hunk 2]
```diff
@@ -1831,11 +1821,7 @@ public async Task FieldIsRead_PartialClass_DifferentFile()
     public int M() => _goo;
 }";
 
-            await new VerifyCS.Test
-            {
-                TestState = { Sources = { source1, source2 } },
-                FixedState = { Sources = { source1, source2 } },
-            }.RunAsync();
+            await VerifyCS.VerifyCodeFixAsync(sources: (source1, source2), fixedSources: (source1, source2), numberOfFixAllIterations: 0);
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs]: > numberOfFixAllIterations: 0

This is the default.

File: src/EditorFeatures/TestUtilities/CodeActions/CSharpCodeFixVerifier`2.cs [hunk 3]
```diff
@@ -72,5 +74,48 @@ public static async Task VerifyCodeFixAsync(string source, DiagnosticResult[] ex
...
+        public static async Task VerifyCodeFixAsync((string first, string second) sources, (string first, string second) fixedSources, int numberOfFixAllIterations)
+            => await VerifyCodeFixAsync(sources, fixedSources, DiagnosticResult.EmptyDiagnosticResults, numberOfFixAllIterations);
+
+        public static async Task VerifyCodeFixAsync((string first, string second) sources, (string first, string second) fixedSources, DiagnosticResult[] expected, int numberOfFixAllIterations)
+        {
+            var test = new Test
+            {
+                TestState = { Sources = { sources.first, sources.second } },
+                FixedState = { Sources = { fixedSources.first, fixedSources.second } },
+                NumberOfFixAllInDocumentIterations = numberOfFixAllIterations
```
Comment (sharwell) [inline src/EditorFeatures/TestUtilities/CodeActions/CSharpCodeFixVerifier`2.cs]: ❗️ `NumberOfFixAllInDocumentIterations` is not the same as `NumberOfFixAllIterations`. We should not confuse the naming.

File: src/EditorFeatures/TestUtilities/CodeActions/CSharpCodeFixVerifier`2.cs [hunk 4]
```diff
@@ -72,5 +74,48 @@ public static async Task VerifyCodeFixAsync(string source, DiagnosticResult[] ex
...
+            test.ExpectedDiagnostics.AddRange(expected);
+            await test.RunAsync();
+        }
+
+        /// <summary>
+        /// Verifies fix for a single diagnostic and iterative batch fix for all diagnostics.
+        /// </summary>
+        /// <param name="source">Original source with multiple diagnostics.</param>
+        /// <param name="fixedSource">Fixed source with code fix applied for a single diagnostic selected by <paramref name="diagnosticSelector"/>.</param>
+        /// <param name="batchFixedSource">Fixed source for iterative batch fix for all diagnostics.</param>
```
Comment (sharwell) [inline src/EditorFeatures/TestUtilities/CodeActions/CSharpCodeFixVerifier`2.cs]: 📝 This is used for all four fix approaches (iterative, fix all in document/project/solution).

File: src/EditorFeatures/VisualBasicTest/RemoveUnusedMembers/RemoveUnusedMembersTests.vb (hunk 1 above)
Comment (sharwell) [reply src/EditorFeatures/VisualBasicTest/RemoveUnusedMembers/RemoveUnusedMembersTests.vb]: The behavior occurred during a Fix All in Document operation. Are you sure the test is not behaving the same way as the IDE?

File: src/EditorFeatures/TestUtilities/CodeActions/CSharpCodeFixVerifier`2.cs [hunk 5]
```diff
@@ -72,5 +74,48 @@ public static async Task VerifyCodeFixAsync(string source, DiagnosticResult[] ex
...
+        /// <param name="diagnosticSelector">Delegate to select a single diagnostic to apply fix to verify <paramref name="fixedSource"/>.</param>
+        public static async Task VerifyFixOneAndFixBatchAsync(
```
Comment (sharwell) [inline src/EditorFeatures/TestUtilities/CodeActions/CSharpCodeFixVerifier`2.cs]: 📝 So far this is really an edge case. I'd like to continue without the overload for a bit to see how things are shaping up.

File: src/EditorFeatures/TestUtilities/CodeActions/CSharpCodeFixVerifier`2.cs [hunk 6]
```diff
@@ -72,5 +74,48 @@ public static async Task VerifyCodeFixAsync(string source, DiagnosticResult[] ex
             test.ExpectedDiagnostics.AddRange(expected);
             await test.RunAsync();
         }
+
+        public static async Task VerifyCodeFixAsync((string first, string second) sources, (string first, string second) fixedSources, int numberOfFixAllIterations)
+            => await VerifyCodeFixAsync(sources, fixedSources, DiagnosticResult.EmptyDiagnosticResults, numberOfFixAllIterations);
+
+        public static async Task VerifyCodeFixAsync((string first, string second) sources, (string first, string second) fixedSources, DiagnosticResult[] expected, int numberOfFixAllIterations)
```
Comment (sharwell) [inline src/EditorFeatures/TestUtilities/CodeActions/CSharpCodeFixVerifier`2.cs]: 📝 The general approach to overloads for these verifier methods is avoid them unless they provide overwhelming benefit without being prone to combinatorial explosions. I don't believe the multiple sources scenario fits for this, and the `numberOfFixAllIterations` certainly wouldn't (it's only used for bugs).

File: src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs (hunk 2 above)
Comment (sharwell) [reply src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs]: The test framework limitation is https://github.com/dotnet/roslyn-sdk/issues/147

File: src/EditorFeatures/VisualBasicTest/RemoveUnusedMembers/RemoveUnusedMembersTests.vb (hunk 1 above)
Comment (mavasani) [reply src/EditorFeatures/VisualBasicTest/RemoveUnusedMembers/RemoveUnusedMembersTests.vb]: My question is I am unable to find any result baseline which gets the test to pass :-)

File: src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs (hunk 2 above)
Comment (mavasani) [reply src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs]: I believe we have large number of IDE tests which have multiple inputs as the documents. Given the test framework does not support the XML markup based baselines, I think we should add helpers to enable easily porting such tests.

File: src/EditorFeatures/TestUtilities/CodeActions/CSharpCodeFixVerifier`2.cs (hunk 3 above)
Comment (mavasani) [reply src/EditorFeatures/TestUtilities/CodeActions/CSharpCodeFixVerifier`2.cs]: Sure, I can rename the parameter.

File: src/EditorFeatures/TestUtilities/CodeActions/CSharpCodeFixVerifier`2.cs (hunk 4 above)
Comment (mavasani) [reply src/EditorFeatures/TestUtilities/CodeActions/CSharpCodeFixVerifier`2.cs]: Will adjust the doc comment.

File: src/EditorFeatures/TestUtilities/CodeActions/CSharpCodeFixVerifier`2.cs (hunk 5 above)
Comment (mavasani) [reply src/EditorFeatures/TestUtilities/CodeActions/CSharpCodeFixVerifier`2.cs]: I don't know if it is really an edge case to have tests which have multiple diagnostics and want to validate a fix on only one of them. I found it very hard to read or write the code that is within this method. In fact, I don't even know how to write it in VB as the only approach here seems to be highly nested object initializers - I don't know if VB also supports it or not. Can you paste a VB snippet that can accomplish the same functionality as this helper?

File: src/EditorFeatures/TestUtilities/CodeActions/CSharpCodeFixVerifier`2.cs (hunk 6 above)
Comment (mavasani) [reply src/EditorFeatures/TestUtilities/CodeActions/CSharpCodeFixVerifier`2.cs]: Can you suggest how to write this in VB?

**Comment by mavasani** [comment]
Stale PR

**Pull request closed without merging**